iroh-net/src/net/udp.rs

```diff
@@ -0,0 +1,150 @@
+            .with_context(|| format!("{addr:?}"))
+    }
+
+    /// Bind to any provided [`SocketAddr`]. Does not prepare for using the socket as QUIC socket.
+    pub async fn bind_full(addr: impl Into<SocketAddr>) -> Result<Self> {
+        Self::bind_raw(addr, false).await
+    }
+
+    async fn bind_raw(addr: impl Into<SocketAddr>, prepare_for_quinn: bool) -> Result<Self> {
+        let addr = addr.into();
+        let network = Network::from(addr.ip());
+        let socket = socket2::Socket::new(
+            network.into(),
+            socket2::Type::DGRAM,
+            Some(socket2::Protocol::UDP),
+        )?;
+
+        if let Err(err) = socket.set_recv_buffer_size(SOCKET_BUFFER_SIZE) {
+            warn!(
+                "failed to set recv_buffer_size to {}: {:?}",
+                SOCKET_BUFFER_SIZE, err
+            );
+        }
+        if let Err(err) = socket.set_send_buffer_size(SOCKET_BUFFER_SIZE) {
+            warn!(
+                "failed to set send_buffer_size to {}: {:?}",
+                SOCKET_BUFFER_SIZE, err
+            );
+        }
+        if network == Network::Ipv6 {
+            // Avoid dualstack
+            socket.set_only_v6(true)?;
+        }
+
+        if prepare_for_quinn {
+            quinn_udp::UdpSocketState::configure((&socket).into())?;
+            // disable nonblocking to ensure socket2 bind works
+            socket.set_nonblocking(false)?;
+        }
+
+        socket.bind(&addr.into())?;
+
+        // Reenable nonblocking
+        socket.set_nonblocking(true)?;
+
+        let socket: std::net::UdpSocket = socket.into();
+
+        // Convert into tokio UdpSocket
+        let socket = tokio::net::UdpSocket::from_std(socket)?;
+
+        if addr.port() != 0 {
+            let local_addr = socket.local_addr()?;
+            ensure!(
+                local_addr.port() == addr.port(),
+                "wrong port bound: {:?}: wanted: {} got {}",
+                network,
+                addr.port(),
+                local_addr.port(),
+            );
+        }
+        Ok(UdpSocket(Some(socket)))
+    }
+}
+
+#[cfg(unix)]
+impl std::os::fd::AsFd for UdpSocket {
+    fn as_fd(&self) -> std::os::fd::BorrowedFd<'_> {
+        self.0.as_ref().expect("not dropped").as_fd()
+    }
+}
+
+#[cfg(windows)]
+impl std::os::windows::io::AsSocket for UdpSocket {
+    fn as_socket(&self) -> std::os::windows::io::BorrowedSocket<'_> {
+        self.0.as_ref().expect("not dropped").as_socket()
+    }
+}
+
+impl From<tokio::net::UdpSocket> for UdpSocket {
+    fn from(socket: tokio::net::UdpSocket) -> Self {
+        Self(Some(socket))
+    }
+}
+
+impl std::ops::Deref for UdpSocket {
+    type Target = tokio::net::UdpSocket;
+
+    fn deref(&self) -> &Self::Target {
+        self.0.as_ref().expect("only removed on drop")
+    }
+}
+
+impl Drop for UdpSocket {
+    fn drop(&mut self) {
+        let std_sock = self.0.take().expect("not yet dropped").into_std();
+        tokio::task::spawn_blocking(move || {
```

Comment:
this panics if it's being dropped outside of a tokio runtime.  That's... scary.

Could we explicitly use `tokio::runtime::Handle::try_current` and do a sync drop if not in a runtime?  Or would that defeat the point?

Reply:
ah yes, I should do that

Reply:
fixed